### ISSUE TITLE
Overhaul @IntegrationTest's TestExecutionListener configuration to align with Spring Framework 4.1

### ISSUE THREAD

**Issue by sbrannen** [spring-projects/spring-boot#1956]
## Status Quo

The implementation of `@IntegrationTest` takes a decidedly restrictive approach to `TestExecutionListener`s:

1. `@IntegrationTest` omits the `SqlScriptsTestExecutionListener` that was introduced in Spring Framework 4.1.
1. The hard-coded configuration for `TestExecutionListener`s in `@IntegrationTest` precludes the ability for developers and third-party frameworks to register custom `TestExecutionListener`s to be used in conjunction with Spring Boot's `IntegrationTestPropertiesListener`.

## Proposal

At the very least, `SqlScriptsTestExecutionListener` should be included in the set of hard-coded listeners for `@IntegrationTest`.

In order to support the registration of custom `TestExecutionListener`s, Spring Boot could adopt the support for ["Automatic discovery of default TestExecutionListeners"](http://docs.spring.io/spring/docs/current/spring-framework-reference/html/testing.html#testcontext-tel-config-automatic-discovery) that was introduced in Spring Framework 4.1. For example, Spring Security already makes use of this feature.

If Spring Boot's `IntegrationTestPropertiesListener` is to be picked up via auto-detection, it should override `AbstractTestExecutionListener.getOrder()` with an appropriate value. See ["Ordering TestExecutionListeners"](http://docs.spring.io/spring/docs/current/spring-framework-reference/html/testing.html#testcontext-tel-config-ordering) for further details.

In general, any `TestExecutionListener` should now (as of Spring Framework 4.1) correctly implement `Ordered` for proper behavior when `@TestExecutionListeners(mergeMode=...)` is set to `MergeMode.MERGE_WITH_DEFAULTS`. See ["Merging TestExecutionListeners"](http://docs.spring.io/spring/docs/current/spring-framework-reference/html/testing.html#testcontext-tel-config-merging) for details.

For greater flexibility, you may also want to consider making `IntegrationTestPropertiesListener` a `public` class.

**Comment by philwebb** [comment]
@sbrannen I started to take a look at this to see if it would solve #2135. Unfortunately I've hit a slight snag. Most `@IntegrationTest` tests also use `@WebAppConfiguration` which declares `@BootstrapWith(WebTestContextBootstrapper.class)`. This means that any `@BootstrapWith` on `@IntegrationTest` is ignored.

I think `BootstrapUtils.resolveTestContextBootstrapper` might need to be updated to deal with multiple `@BootstrapWith` annotations. Perhaps they could be `@Ordered`?

**Comment by philwebb** [comment]
I've added a new `@WebIntegrationTest` annotation which works as long as you don't also use `@WebAppConfiguration`. We could probably deprecate `@IntegrationTest` in 1.3.x since it doesn't offer much over `@TestPropertySource` when you are not running a web test. We should probably still find a way to protect against accidentally using both `@WebAppConfiguration` and `@WebIntegrationTest`.

**Comment by sbrannen** [comment]
> I've added a new `@WebIntegrationTest` annotation which works as long as you don't also use `@WebAppConfiguration`.

`@WebIntegrationTest` seems like a pretty elegant route to take, and yes... `@WebIntegrationTest` should not be used in conjunction with `@WebAppConfiguration`.

To address this, Spring Boot could write a _validation_ check in `SpringApplicationContextLoader` that verifies that `@WebAppConfiguration` is only present on `@WebIntegrationTest` and nowhere else, throwing an `IllegalStateException` if the assertion fails.

> We could probably deprecate `@IntegrationTest` in 1.3.x since it doesn't offer much over 
`@TestPropertySource` when you are not running a web test.

That's only partially true: based on the in-line comments, `@IntegrationTest` properties go before external config and after system; whereas, properties configured via `@TestPropertySource` always have the highest precedence. So dropping support for `@IntegrationTest` semantics would likely break backwards compatibility for some users of Spring Boot.

> We should probably still find a way to protect against accidentally using both `@WebAppConfiguration` and `@WebIntegrationTest`.

See my suggestion above. ;)

**Comment by sbrannen** [comment]
> I think `BootstrapUtils.resolveTestContextBootstrapper` might need to be updated to deal with multiple `@BootstrapWith` annotations. Perhaps they could be `@Ordered`?

Hmmmm... I'm not sure that we want to go that route: it's not foreseen that two bootstrappers could be used simultaneously.

How would you image that working in practice?

Or are you simply suggesting that a particular bootstrapper take precedence and override any other declarations?

**Comment by philwebb** [comment]
@sbrannen The additional validation check in `SpringApplicationContextLoader` will help but currently if you accidentally use more than one `@BootstrapWith` annotation it's somewhat arbitrary as to which one gets used. If you shouldn't use more than one bootstrappers perhaps there should be an explicit failure if you try.

**Comment by sbrannen** [comment]
> If you shouldn't use more than one bootstrappers perhaps there should be an explicit failure if you try.

Yeah, that sounds like something we should consider adding to `spring-test`.

Can you please open a JIRA issue for that?

**Comment by sbrannen** [comment]
Thanks for creating the [JIRA](https://jira.spring.io/browse/SPR-12602) issue!

**Comment by nealeu** [comment]
Just picked up on this.  We use Liquibase with a selected Liquibase contexts specified to initialise our schema for integration tests.
How would the SqlScriptsTestExecutionListener being hard-coded compare for us?

**Comment by sbrannen** [comment]
`SqlScriptsTestExecutionListener` is already registered by default in `spring-test` (i.e., core Spring Framework), and it doesn't do anything unless you annotate your test class or test method with `@Sql`.

So there are no adverse effects of having `SqlScriptsTestExecutionListener` registered by default with Spring Boot's `@IntegrationTest` annotation. Having it there just brings Spring Boot up to date with the latest features of Spring's core testing support.

In other words, it just provides developers with more options. ;)

**Comment by kiview** [comment]
When using `@IntegrationTest`, files inside the normal resource are accessible by the application (are inside the classpath?). However, when using `@WebIntegrationTest`, these files are not accessible, only files inside a resource directory inside the test directory are accessible. Is this behavior intended or some side effect?

**Comment by sbrannen** [comment]
@kiview, could you please create an example project that reproduces the behavior you are experiencing regarding classpath resources, `@IntegrationTest`, and `@WebIntegrationTest`?

**Comment by sbrannen** [comment]
Since Spring Boot depends on Spring Framework 4.2, the `prepareTestInstance()` method in `IntegrationTestPropertiesListener` can now be simplified as follows (i.e., there's no longer a need to work with `AnnotationAttributes` unless you just want to):

```java
public void prepareTestInstance(TestContext testContext) throws Exception {
	IntegrationTest integrationTest = AnnotatedElementUtils.
			getMergedAnnotation(testContext.getTestClass(), IntegrationTest.class);
	if (integrationTest != null) {
		addPropertySourceProperties(testContext, integrationTest.value());
	}
}
```

**Comment by sbrannen** [comment]
Just brainstorming here...

If Spring Boot deletes the declaration of `@TestExecutionListeners` on `@IntegrationTest` and includes `IntegrationTestPropertiesListener` in `spring.factories`, then the `IntegrationTestPropertiesListener` could conditionally (based on the presence of `@IntegrationTest` or `@WebIntegrationTest`)  _instruct_ the `ServletTestExecutionListener` not to create mocks by setting the `ServletTestExecutionListener.POPULATED_REQUEST_CONTEXT_HOLDER_ATTRIBUTE` in the `TestContext` to `Boolean.TRUE`.

Doing so would effectively disable the `ServletTestExecutionListener`, and it would allow Spring Boot to adopt the support for "_Automatic discovery of default TestExecutionListeners_" that was introduced in Spring Framework 4.1, thereby avoiding many of the shortcomings of the current implementation.

Thoughts?

**Comment by sbrannen** [comment]
Relates to #2505.

**Comment by philwebb** [comment]
I think that #5230 has effectively fixed this issue since `@IntegrationTest` now uses `@BootstrapWith`
